sdk/src/sdk/vdb/grpc/kuksa_val_v2/Metadata.cpp

```diff
@@ -0,0 +1,414 @@
+        } catch (...) {
+            logger().error("Unknown exception catched while notifying metadata query error");
+        }
+    }
+
+private:
+    MetadataList_t                           m_cachedMetadata;
+    std::set<std::string>                    m_missingSignals;
+    std::function<void(MetadataList_t&&)>    m_successCallback;
+    std::function<void(const grpc::Status&)> m_errorCallback;
+};
+
+void notifyQueryInitiators(std::deque<Query>&& queries) {
+    for (auto& query : queries) {
+        query.notifyInitiator();
+    }
+}
+
+void notifyQueryInitiators(std::deque<Query>&& queries, const grpc::Status& status) {
+    for (auto& query : queries) {
+        query.notifyInitiator(status);
+    }
+}
+
+} // namespace
+
+class MetadataAgentImpl : public MetadataAgent {
+public:
+    explicit MetadataAgentImpl(std::shared_ptr<BrokerAsyncGrpcFacade> asyncBrokerFacade)
+        : m_asyncBrokerFacade(std::move(asyncBrokerFacade)) {}
+
+    void query(const SignalNameList_t&                    signalNames,
+               std::function<void(MetadataList_t&&)>&&    onSuccess,
+               std::function<void(const grpc::Status&)>&& onError) override;
+    void invalidate(grpc::StatusCode statusCode) override;
+
+    [[nodiscard]] MetadataPtr_t getByNumericId(numeric_id_t numericId) const override {
+        std::shared_lock lock(m_mutex);
+        return m_cache.getById(numericId);
+    }
+
+private:
+    void              addCachedMetadata(Query& query, const SignalNameList_t& signalNames);
+    void              addQuery(Query&& query);
+    void              addSignalsToRequestQueue(const std::set<std::string>& signals);
+    void              addSignalToRequestQueue(const std::string& signalName);
+    bool              isSignalPartOfActiveRequests(const std::string& signalName) const;
+    void              triggerMetadataRequests();
+    void              updateActiveRequests(const std::shared_ptr<Request>& request);
+    std::deque<Query> updateQueriesAndExtractFulfilled(const MetadataPtr_t& metadata);
+    std::deque<Query> extractAffectedQueries(const std::string& signalName);
+    void              cancelActiveRequests();
+
+    std::shared_ptr<BrokerAsyncGrpcFacade> m_asyncBrokerFacade;
+
+    mutable std::shared_mutex          m_mutex;
+    MetadataCache                      m_cache;
+    std::deque<Query>                  m_pendingQueries;
+    std::deque<std::string>            m_pendingSignals;
+    std::set<std::shared_ptr<Request>> m_activeRequests;
+};
+
+std::shared_ptr<MetadataAgent>
+MetadataAgent::create(const std::shared_ptr<BrokerAsyncGrpcFacade>& brokerFacade) {
+    return std::make_shared<MetadataAgentImpl>(brokerFacade);
+}
+
+void MetadataAgentImpl::cancelActiveRequests() {
+    for (const auto& request : m_activeRequests) {
+        request->cancel();
+    }
+}
+
+void MetadataAgentImpl::invalidate(grpc::StatusCode statusCode) {
```

Comment:
should this not only be called in one scenario when the databroker is gone and restarts so only for `Status::UNAVAILABLE` right? Why do we need to handover status code here?

Reply:
I wanted to be flexible here ...